### PR TITLE
Add Model.readyPromise and Primitive.readyPromise

### PR DESCRIPTION
**Pull request by mramato**
We should also deprecate `Model.readyToRender`, but I wanted to make sure @pjcozzi was okay with it first.  Having an event that only fires once are just the sort of problem promises are meant to solve.

### CODE REVIEW

File: Source/Scene/Primitive.js [hunk 1]
```diff
@@ -798,6 +811,7 @@ define([
                     }, function(error) {
                         that._error = error;
                         that._state = PrimitiveState.FAILED;
+                        this._readyPromise.resolve(this);
```
Comment (pjcozzi) [inline Source/Scene/Primitive.js]: Throughout `Primitive.js`, can we resolve this in `frameState.afterRender` so the called function can not interfere with the render loop?  This is done in `Model.js`.  If we do this a lot, perhaps we'll introduce something like `frameState.resolveAfterRender(promise)`.

File: Specs/Scene/ModelSpec.js [hunk 2]
```diff
@@ -150,6 +152,12 @@ defineSuite([
         verifyRender(duckModel);
     });
 
+    it('resolves readyPromise', function() {
+        waitsForPromise(duckModel.readyPromise, function(model) {
+            verifyRender(model);
```
Comment (pjcozzi) [inline Specs/Scene/ModelSpec.js]: I know it is already implicit, but perhaps we should explicitly validate `model.ready`.

File: Source/Scene/Model.js [hunk 3]
```diff
@@ -564,6 +567,18 @@ define([
         },
 
         /**
+         * Gets a promise that resolves when the model is ready to render.
```
Comment (pjcozzi) [inline Source/Scene/Model.js]: Can we make this doc as good as the doc for `this.readyToRender` and then change all doc links to `readyToRender` to this?

**Comment by pjcozzi** [comment]
Can we update the 3D model Sandcastle example to use this?  It will still be worthwhile even with the entity API.

**Comment by pjcozzi** [comment]
> We should also deprecate `Model.readyToRender`.

Agreed.  Given that I expect it is widely used, let's deprecate now and remove in 1.8 or 1.9.

**Comment by pjcozzi** [comment]
Also, a nice short tech blog post (for JavaScript in general) may be when to use raw callbacks vs. events vs. promises.  Submit an issue to the cesium website repo if you think so.

File: Source/Scene/Primitive.js (hunk 1 above)
Comment (mramato) [inline Source/Scene/Primitive.js]: That's a good idea, I ran into a similar problem recently with the zoom stuff that caused be to insert an extra render in Viewer code; I think using `afterRender` there will fix that too. Thanks for the heads up.

File: Source/Scene/Model.js (hunk 3 above)
Comment (mramato) [inline Source/Scene/Model.js]: Definitely, I was just waiting for the okay from you to deprecate.

**Comment by mramato** [comment]
> Can we update the 3D model Sandcastle example to use this? It will still be worthwhile even with the entity API.

Sure, but the ported entity example doesn't need it since viewer.zoomTo will take care of it under the hood.

> Also, a nice short tech blog post (for JavaScript in general) may be when to use raw callbacks vs. events vs. promises. Submit an issue to the cesium website repo if you think so.

Maybe, but there are already some excellent articles out there, [thie one for example](http://www.html5rocks.com/en/tutorials/es6/promises/).  I think it would be better to just have a paragraph or two as part of a larger tutorial on the high-level architecture/idioms of Cesium.  I have a plan in my head for this, I'll write up an issue for it on the website.

File: Specs/Scene/ModelSpec.js (hunk 2 above)
Comment (mramato) [inline Specs/Scene/ModelSpec.js]: I added one at the start of `verifyRender` just to cover all of our bases.

**Comment by mramato** [comment]
Okay, this is ready. I ended up going down the :rabbit2: hole a bit and uncovered a few issues with how we handle worker exceptions as well as found out we have some untested paths in `Primitive`. This change ended up being bigger than expected because of that, but they are all good fixes for problems we didn't know we had.

**Comment by pjcozzi** [comment]
Looks good.

**Pull request merged**